File: Data Structures & Algorithms/products-of-array-discluding-self/submission-9.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
      total_length = len(nums)
      left = [0] * total_length
      right = [0] * total_length
      result = [0] * total_length

      value = 1
      
      for index in range(total_length):
         left[index] = value
         value *= nums[index]

      print(left)

      value = 1

      for index in reversed(range(total_length)):
         right[index] = value
         value *= nums[index]

      for index, num in enumerate(nums):
         result[index] = left[index] * right[index]
      
      return result

File: Data Structures & Algorithms/subsets/submission-9.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        dfs(nums, new ArrayList<>(), result, 0);
        return result;
    }

    public void dfs(int [] nums, List<Integer> current, List<List<Integer>> result, int index){
        if(index >= nums.length){
            result.add(new ArrayList<>(current));
            return;
        }

        current.add(nums[index]);
        dfs(nums, current, result, index + 1);
        current.remove(current.size() - 1);
        dfs(nums, current, result, index + 1);
    }
}
